File: app/workflow/transformers/environment_generator.rb
# frozen_string_literal: true

require_relative 'base'

module Workflow
  module Transformers
    # Clones the isolated source_env configurations and bootstraps all target environment configurations
    class EnvironmentGenerator < Base
      def call(workspace)
        source_env = workspace.source_env
        overlay_files = workspace.source_overlay_files

        workspace.target_envs.each do |env|
          overlay_files.each do |virtual_path, content|
            new_path = virtual_path.sub("overlay/#{source_env}", "overlay/#{env}")
            
            # Deep clone and replace string instances of source_env with env
            workspace.manifests[new_path] = deep_replace(content, source_env, env)
          end
        end

        workspace
      end

      private

      def deep_replace(node, source, target)
        case node
        when Hash
          node.transform_values { |v| deep_replace(v, source, target) }
        when Array
          node.map { |v| deep_replace(v, source, target) }
        when String
          node.gsub(source, target)
        else
          node
        end
      end
    end
  end
end

File: internal/transformers/environment_generator.go
// Package transformers is the pure, side-effect-free part of the sync
// pipeline: every function here takes a *manifest.Workspace and returns it
// mutated, with no I/O and no dependency on anything outside its arguments.
// That purity is exactly what lets a Temporal workflow call these directly
// instead of wrapping them in activities - see internal/workflows/syncworkloads.go.
package transformers

import (
	"slices"
	"strings"

	"github.com/abradner/workflow/internal/manifest"
)

// EnvironmentGenerator clones the source environment's overlay files into
// every other target environment, rewriting embedded env-name references as
// it goes. It must run first in the pipeline - everything downstream expects
// the full set of target-env overlays to already exist.
func EnvironmentGenerator(ws *manifest.Workspace) *manifest.Workspace {
	sourceEnv := ws.SourceEnv
	overlayFiles := ws.SourceOverlayFiles()

	for _, env := range ws.TargetEnvs {
		if env == sourceEnv {
			continue
		}

		for virtualPath, content := range overlayFiles {
			newPath := strings.Replace(virtualPath, "overlay/"+sourceEnv, "overlay/"+env, 1)
			ws.Manifests[newPath] = deepReplace(content, sourceEnv, env)
		}
	}

	// Drop the source environment's own files unless it was explicitly
	// requested as one of the deployment targets.
	if !slices.Contains(ws.TargetEnvs, sourceEnv) {
		prefix := "overlay/" + sourceEnv + "/"
		for path := range ws.Manifests {
			if strings.HasPrefix(path, prefix) {
				delete(ws.Manifests, path)
			}
		}
	}

	return ws
}

// deepReplace walks a parsed-YAML value tree, returning a deep copy with
// every occurrence of source replaced by target in every string leaf.
func deepReplace(node any, source, target string) any {
	switch v := node.(type) {
	case map[string]any:
		out := make(map[string]any, len(v))
		for k, val := range v {
			out[k] = deepReplace(val, source, target)
		}
		return out
	case []any:
		out := make([]any, len(v))
		for i, val := range v {
			out[i] = deepReplace(val, source, target)
		}
		return out
	case string:
		return strings.ReplaceAll(v, source, target)
	default:
		return node
	}
}
